services/collectd/service.go

```diff
@@ -336,32 +358,49 @@ func (s *Service) serve() {
 	//   Ethernet).
 	buffer := make([]byte, 1452)
 
+	type ReadPayload struct {
+		n   int   // bytes read by ReadFromUDP
+		err error // any error returend by ReadFromUDP
+	}
+
+	rp := ReadPayload{}
+
+	reads := make(chan ReadPayload)
+
+	go func() {
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			default:
+				rp.n, _, rp.err = s.conn.ReadFromUDP(buffer)
+				reads <- rp
+			}
+		}
+	}()
+
 	for {
 		select {
-		case <-s.done:
-			// We closed the connection, time to go.
+		case <-ctx.Done():
 			return
-		default:
-			// Keep processing.
-		}
-
-		n, _, err := s.conn.ReadFromUDP(buffer)
-		if err != nil {
-			if strings.Contains(err.Error(), "use of closed network connection") {
-				select {
-				case <-s.done:
-					return
-				default:
-					// The socket wasn't closed by us so consider it an error.
+		case cur := <-reads:
+			if cur.err != nil {
+				if strings.Contains(rp.err.Error(), "use of closed network connection") {
```

Comment:
this method of checking for an error is unfortunate. I don't know what would be better though in this case.

Reply:
ya -- i don't like it.  i'm mostly preserving how it was done previously.